common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -877,7 +886,7 @@ void closeAllFiles() {
       }
```

Comment:
Do we want to truncate data file as well to the length recorded by the last offset in the index file?
https://github.com/apache/spark/blob/dfa3978d9191e02eabf65d1829c970644d25d57e/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java#L869-L871

BTW, just realized this limitation in GitHub (https://github.com/isaacs/github/issues/284).
People know any workaround?

Reply:
Truncating would be good, it will keep the file consistent with the index files as well.

Reply:
What do we do when the truncate of data file here fails?

Reply:
Also we truncate the data file before `closeAllFiles` is invoked in `finalizeShuffleMerge`
```
 partition.dataChannel.truncate(partition.getPosition());
```

Reply:
That truncation is to truncate to the last successfully appended block.
However, if the index/meta files are not in sync, we would have these blocks in the end of the file that won't be consumed.

Reply:
Yes, but I think truncating the data file twice while finalizing is unnecessary. Also we can't simply ignore the exceptions while truncating all these 3 files during close. I think truncation of these files should be separated from `closeAllFiles`.
This is the change I have made:
```
  public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOException {
     ...
      while (partitionsIter.hasNext()) {
        AppShufflePartitionInfo partition = partitionsIter.next();
        synchronized (partition) {
          try {
            partition.finalizePartition();
            bitmaps[idx] = partition.mapTracker;
            reduceIds[idx] = partition.reduceId;
            sizes[idx++] = partition.getPosition();
          } catch (IOException ioe) {
            logger.warn("Exception while finalizing shuffle partition {} {} {}", msg.appId,
              msg.shuffleId, partition.reduceId, ioe);
          } finally {
            partition.closeAllFiles();
            // The partition should be removed after the files are written so that any new stream
            // for the same reduce partition will see that the data file exists.
            partitionsIter.remove();
          }
        }
      }
      mergeStatuses = new MergeStatuses(msg.shuffleId, bitmaps, reduceIds, sizes);
    }
    partitions.remove(appShuffleId);
    logger.info("Finalized shuffle {} from Application {}.", msg.shuffleId, msg.appId);
    return mergeStatuses;
  }

   private void finalizePartition() throws IOException {
      if (position != lastChunkOffset) {
        try {
          updateChunkInfo(position, lastMergedMapIndex);
        } catch (IOException ioe) {
          if (seekFailed) {
            // If seek has failed then fail finalize of this partition by propagating the exception.
            throw ioe;
          }
          // If seek hasn't failed then ignore this exception.
        }
      }
      // Get rid of any partial block data at the end of the file. This could either
      // be due to failure or a request still being processed when the shuffle
      // merge gets finalized.
      dataChannel.truncate(lastChunkOffset);
      metaFile.setLength(metaFile.getFilePointer());
      indexFile.setLength(indexFile.getFilePointer());
    }
```
Any exception while truncating data/meta/index file during finalize will propagate the exception to the caller and that partition will not be considered merged.